src/opossum_lib/tree_generation.py

```diff
@@ -0,0 +1,88 @@
+    if not graph.edges():
+        return graph
+
+    if is_weakly_connected(graph):
+        tree = tree or DiGraph()
+        edges_bfs = edge_bfs(graph, source)
+        visited_edges = []
+        for edge in edges_bfs:
+            _add_source_node(edge, graph, tree)
+            _add_target_node_and_edge(edge, graph, tree)
+            visited_edges += [edge]
+        # check if there is a directed path from source to each element node
+        # if not, we need to construct sub-graphs from the unreached edges
+        # and construct trees for all of these sub-graphs, the result
+        # will be an unconnected graph
+        unreached_edges = set(graph.edges).difference(set(visited_edges))
+        if unreached_edges:
+            induced_subgraph = graph.edge_subgraph(unreached_edges)
+            source = _get_node_without_incoming_edge(induced_subgraph)
+            tree_component = generate_tree_from_graph(induced_subgraph, source, tree)
+            tree.add_nodes_from(tree_component.nodes(data=True))
+            tree.add_edges_from(tree_component.edges(data=True))
+
+    else:  # get connected components
+        tree = DiGraph()
+        for connected_set in weakly_connected_components(
+            graph
+        ):  # returns only a set of nodes without edges
+            connected_subgraph = graph.subgraph(connected_set).copy()
+            # if the documents node is not in the subgraph we choose
+            # any elements node without incoming edge
+            source = (
+                "SPDXRef-DOCUMENT"
+                if "SPDXRef-DOCUMENT" in connected_set
+                else _get_node_without_incoming_edge(connected_subgraph)
+            )
+            tree_component = generate_tree_from_graph(connected_subgraph, source)
+            tree.add_nodes_from(tree_component.nodes(data=True))
+            tree.add_edges_from(tree_component.edges(data=True))
+
+    return tree
+
+
+def _add_target_node_and_edge(
+    edge: Tuple[str, str], graph: DiGraph, tree: DiGraph
+) -> None:
+    if edge[1] not in tree:
+        target_node_data = graph.nodes[edge[1]]
+        tree.add_node(edge[1], **target_node_data)
+        tree.add_edge(*edge)
+    else:
+        # if the child node is already in the graph duplicate the node by adding the
+        # parent node as prefix
+        target_node_data = graph.nodes[edge[1]]
+        duplicated_node = edge[0] + "_" + edge[1]
+        tree.add_node(duplicated_node, **target_node_data)
+        tree.add_edge(edge[0], duplicated_node)
+
+
+def _add_source_node(edge: Tuple[str, str], graph: DiGraph, tree: DiGraph) -> None:
+    if edge[0] not in tree:
+        source_node_data: Dict[str, Union[CreationInfo, Package, File]] = graph.nodes[
+            edge[0]
+        ]
+        tree.add_node(edge[0], **source_node_data)
+
+
+def _get_node_without_incoming_edge(graph: DiGraph) -> Any:
+    for node, degree in graph.in_degree():
+        if degree == 0 and "element" in graph.nodes[node]:
```

Comment:
What is the `"element" in graph.nodes[node]` condition doing? Might be good to extract that into a function with a meaningful name (returning a boolean) to make the code more human-readable.
Might also be good to extract the string "element" into a constant, that would also help understanding what it is used for.